### DISCUSSION TITLE
I have a problem"Nonlinear solve did not converge due to DIVERGED_FNORM_NAN." The inputfile as follows:

### DISCUSSION THREAD

**Discussion by TAIYONGZOU** [idaholab/moose#26148] [Q&A Modules: Phase field]
#
# KKS ternary (3 chemical component) system example in the split form
# We track c1 and c2 only, since c1 + c2 + c3 = 1
#

[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 150
  ny = 150
  nz = 0
  xmin = -25
  xmax = 25
  ymin = -25
  ymax = 25
  zmin = 0
  zmax = 0
  elem_type = QUAD4
[]

[AuxVariables]
  [./Fglobal]
    order = CONSTANT
    family = MONOMIAL
  [../]
 
  [./bounds_dummy]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[Variables]
  # order parameter
  [./eta]
    order = FIRST
    family = LAGRANGE
  [../]

  # solute 1 concentration
  [./c1]
    order = FIRST
    family = LAGRANGE
  [../]

  # solute 2 concentration
  [./c2]
    order = FIRST
    family = LAGRANGE
  [../]

  [./c3]
    order = FIRST
    family = LAGRANGE
  [../]

  # chemical potential solute 1
  [./w1]
    order = FIRST
    family = LAGRANGE
  [../]

  # chemical potential solute 2
  [./w2]
    order = FIRST
    family = LAGRANGE
  [../]

  [./w3]
    order = FIRST
    family = LAGRANGE
  [../]

  # Liquid phase solute 1 concentration
  [./c1l]
    order = FIRST
    family = LAGRANGE 
  [../]

  # Liquid phase solute 2 concentration
  [./c2l]
    order = FIRST
    family = LAGRANGE
  [../]

  [./c3l]
    order = FIRST
    family = LAGRANGE
  [../]

  # Solid phase solute 1 concentration
  [./c1s]
    order = FIRST
    family = LAGRANGE
  [../]

  # Solid phase solute 2 concentration
  [./c2s]
    order = FIRST
    family = LAGRANGE
  [../]

  [./c3s]
    order = FIRST
    family = LAGRANGE
  [../]

[]

[ICs]
  [./eta]
    type = RndBoundingBoxIC
    x1 = -25.0
    x2 = 0.0
    y1 = -25.0
    y2 = 25.0
    mx_invalue = 1.0
    mn_invalue = 1.0
    mx_outvalue = 1e-5
    mn_outvalue = 1e-5
    variable = eta
  [../]
  [./c1]
    type = RndBoundingBoxIC
    x1 = -25.0
    x2 = 0.0
    y1 = -25.0
    y2 = 25.0
    mx_invalue = 1e-5
    mn_invalue = 1e-5
    mx_outvalue = 0.075
    mn_outvalue = 0.075
    variable = c1
  [../]
  [./c2]
    type = RndBoundingBoxIC
    x1 = -25.0
    x2 = 0.0
    y1 = -25.0
    y2 = 25.0
    mx_invalue = 1e-5
    mn_invalue = 1e-5
    mx_outvalue = 0.02
    mn_outvalue = 0.02
    variable = c2
  [../]
  [./c3]
    type = RndBoundingBoxIC
    x1 = -25.0
    x2 = 0.0
    y1 = -25.0
    y2 = 25.0
    mx_invalue = 0.5
    mn_invalue = 0.5
    mx_outvalue = 0.405
    mn_outvalue = 0.405
    variable = c3
  [../]
[]

[Bounds]
  [./eta_upper_bound]
    type = ConstantBoundsAux
    variable = bounds_dummy
    bounded_variable = eta
    bound_type = upper
    bound_value = 1
  [../]

  [./eta_lower_bound]
    type = ConstantBoundsAux
    variable = bounds_dummy
    bounded_variable = eta
    bound_type = lower
    bound_value =  1e-5
  [../]

  [./c1_upper_bound]
    type = ConstantBoundsAux
    variable = bounds_dummy
    bounded_variable = c1
    bound_type = upper
    bound_value = 0.075
  [../]

  [./c1_lower_bound]
    type = ConstantBoundsAux
    variable = bounds_dummy
    bounded_variable = c1
    bound_type = lower
    bound_value =  1e-5
  [../]

  [./c2_upper_bound]
    type = ConstantBoundsAux
    variable = bounds_dummy
    bounded_variable = c2
    bound_type = upper
    bound_value = 0.02
  [../]

  [./c2_lower_bound]
    type = ConstantBoundsAux
    variable = bounds_dummy
    bounded_variable = c2
    bound_type = lower
    bound_value =  1e-5
  [../]

  [./c3_upper_bound]
    type = ConstantBoundsAux
    variable = bounds_dummy
    bounded_variable = c3
    bound_type = upper
    bound_value = 0.905
  [../]

  [./c3_lower_bound]
    type = ConstantBoundsAux
    variable = bounds_dummy
    bounded_variable = c3
    bound_type = lower
    bound_value =  1e-5
  [../]
[]

[Materials]
  # Free energy of the liquid
  [./fl]
    type = DerivativeParsedMaterial
    property_name = fl
    coupled_variables = 'c1l c2l c3l'
    expression = '6.24150934e+18*1e-27*((c1l*224282047306839)+(c2l*2594493.61)-(26664558343141.2*(c3l))+(11415.122*(c1l*log(c1l)+c2l*log(c2l)+(c3l)*log(c3l)))+(c1l*(c3l)*(-8746.09+8194.99*(c3l-c1l)))-(27240.16*c2l*(c3l))+(c1l*c2l*(-97685.6+14347*(c1l-c2l))-1674*(c1l-c2l)^2-43361*(c1l-c2l)^3)+(c3l)*c1l*c2l*(-127783.6*(c3l)+3520.3*c1l+246657*c2l))'
  [../]

  # Free energy of the solid
  [./fs]
    type = DerivativeParsedMaterial
    property_name = fs
    coupled_variables = 'c1s c2s c3s'
    expression = '6.24150934e+18*1e-27*(-9525723.686*c1s+c2s*90269.85157-516872.9218*(c3s)+11415.122*(c1s*log(c1s)+c2s*log(c2s)+(c3s)*log(c3s))+(c3s)*c1s*(-9654.24+11062.57*(c3s-c1s))-24872.696*(c3s)*c2s-5000*c1s*c2s-5000*(c3s)*c1s*c2s)'
  [../]

  # h(eta)
  [./h_eta]
    type = SwitchingFunctionMaterial
    h_order = HIGH
    eta = eta
  [../]

  # g(eta)
  [./g_eta]
    type = BarrierFunctionMaterial
    g_order = SIMPLE
    eta = eta
  [../]

  # constant properties
  [./constants]
    type = GenericConstantMaterial
    prop_names  = 'M   L   eps_sq'
    prop_values = '0.7 0.7 1.0  '
  [../]
[]

[Kernels]
  # enforce c1 = (1-h(eta))*c1l + h(eta)*c1s
  [./PhaseConc1]
    type = KKSPhaseConcentration
    ca       = c1l
    variable = c1s
    c        = c1
    eta      = eta
  [../]

  # enforce c2 = (1-h(eta))*c2l + h(eta)*c2s
  [./PhaseConc2]
    type = KKSPhaseConcentration
    ca       = c2l
    variable = c2s
    c        = c2
    eta      = eta
  [../]

  [./PhaseConc3]
    type = KKSPhaseConcentration
    ca       = c3l
    variable = c3s
    c        = c3
    eta      = eta
  [../]

  # enforce pointwise equality of chemical potentials
  [./ChemPotSolute1]
    type = KKSPhaseChemicalPotential
    variable = c1l
    cb       = c1s
    fa_name  = fl
    fb_name  = fs
    
  [../]
  [./ChemPotSolute2]
    type = KKSPhaseChemicalPotential
    variable = c2l
    cb       = c2s
    fa_name  = fl
    fb_name  = fs
   
  [../]
  [./ChemPotSolute3]
    type = KKSPhaseChemicalPotential
    variable = c3l
    cb       = c3s
    fa_name  = fl
    fb_name  = fs
    
  [../]

  #
  # Cahn-Hilliard Equations
  #
  [./CHBulk1]
    type = KKSSplitCHCRes
    variable = c1
    ca       = c1l
    fa_name  = fl
    w        = w1
   
  [../]
  [./CHBulk2]
    type = KKSSplitCHCRes
    variable = c2
    ca       = c2l
    fa_name  = fl
    w        = w2
   
  [../]

  [./CHBulk3]
    type = KKSSplitCHCRes
    variable = c3
    ca       = c3l
    fa_name  = fl
    w        = w3
    
  [../]

  [./dc1dt]
    type = CoupledTimeDerivative
    variable = w1
    v = c1
  [../]
  [./dc2dt]
    type = CoupledTimeDerivative
    variable = w2
    v = c2
  [../]
  [./dc3dt]
    type = CoupledTimeDerivative
    variable = w3
    v = c3
  [../]

  [./w1kernel]
    type = SplitCHWRes
    mob_name = M
    variable = w1
  [../]
  [./w2kernel]
    type = SplitCHWRes
    mob_name = M
    variable = w2
  [../]
  [./w3kernel]
    type = SplitCHWRes
    mob_name = M
    variable = w3
  [../]

  #
  # Allen-Cahn Equation
  #
  [./ACBulkF]
    type = KKSACBulkF
    variable = eta
    fa_name  = fl
    fb_name  = fs
    w        = 1.0
    
  [../]
  [./ACBulkC1]
    type = KKSACBulkC
    variable = eta
    ca       = c1l
    cb       = c1s
    fa_name  = fl
    
  [../]
  [./ACBulkC2]
    type = KKSACBulkC
    variable = eta
    ca       = c2l
    cb       = c2s
    fa_name  = fl
    
    [../]
  [./ACBulkC3]
    type = KKSACBulkC
    variable = eta
    ca       = c3l
    cb       = c3s
    fa_name  = fl
    
  [../]
  [./ACInterface]
    type = ACInterface
    variable = eta
    kappa_name = eps_sq
  [../]
  [./detadt]
    type = TimeDerivative
    variable = eta
  [../]
[]

[AuxKernels]
  [./GlobalFreeEnergy]
    variable = Fglobal
    type = KKSGlobalFreeEnergy
    fa_name = fl
    fb_name = fs
    w = 1.0
  [../]
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'

  petsc_options_iname = '-pc_type -sub_pc_type -sub_pc_factor_shift_type'
  petsc_options_value = 'asm      ilu          nonzero'

  l_max_its = 100
  nl_max_its = 100

  num_steps = 50
  dt = 0.1
  
[]

#
# Precondition using handcoded off-diagonal terms
#
[Preconditioning]
  [./full]
    type = SMP
    full = true
  [../]
[]

[Outputs]
  exodus = true
[]

**Comment by GiudGiud** [comment]
Hello

If c3 is simply c1+c2 you should just make it an auxiliary variable. 

That will make the nonlinear system smaller and easier to solve

Guillaume
